We'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. We've made a few modifications from the original dataset that was uploaded to Kaggle:

We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment
We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)


The data dictionary provided with data is as follows:
- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings. You'll also become familiar with some of the unique benefits jupyter notebook provides for pandas.


---



I found the type of encoding for this file by doing as below to assign the type of encoding in read_csv() method since I was getting "unicodedecodeerror: 'utf-8' codec can't decode byte 0xdc in position 23: invalid continuation byte".

In [1]:
file = '/Users/kitaeklee/Desktop/Data/DQ/DQ_ebay_car_sales_data/DQ_ebay_car_sales_data.csv'

import chardet
with open(file, 'rb') as rawdata:
    result = chardet.detect(rawdata.read(100000))
result


{'encoding': 'Windows-1252', 'confidence': 0.73, 'language': ''}

In [2]:
import pandas as pd

autos = pd.read_csv(file, encoding = 'Windows-1252')

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

- There are some columns with null values but none of them have more than 20% of the total number of values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores. Camel case write phrases without spaces of punctuation, indicating the separation of words with a single capitalized letter. Sneak case replace each space by an underscore(__).

---

Therefore, I converted the column names from camel case to sneak case and reword some of the column names based on the data dictionary to be more descriptive as below.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.rename(columns={'dateCrawled' : 'date_crawled',
                      'offerType' : 'offer_type',
                      'vehicleType' : 'vehicle_type',
                      'yearOfRegistration': 'registrationn_year',
                     'monthOfRegistration' : 'registration_month',
                      'fuelType' : 'fuel_type',
                     'notRepairedDamage' : 'unrepaired_damage',
                      'dateCreated' : 'ad_created',
                     'nrOfPictures' : 'nr_of_pictures',
                     'postalCode' : 'postal_code',
                     'lastSeen' : 'last_seen', 
                 'odometer':'odometer_km'},
                 inplace=True)

Look for dupliactes where all or almost all values are the same in text columns. 

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registrationn_year  gearbox       powerPS  model  \
count         44905        50000.000000    47320  50000.000000  47242   
unique            8                 NaN        2           NaN    245   
top       limousine                 NaN  manuell           NaN   golf   
freq          12859                 NaN    36993           NaN   4024   
mean            NaN         2005.073280      NaN    116.355920    NaN   
std             NaN          105.712813      NaN    209.216627    NaN   
min             NaN         1000.000000      NaN      0.000000    NaN   
25%             NaN         1999.000000      NaN     70.000000    NaN   
50%             NaN         2003.000000      NaN    105.000000    NaN   
75%             NaN         2008.000000      NaN    150.000000    NaN   
max             NaN         9999.000000      NaN  17700.000000    NaN   

       odometer_km  registration_month fuel_type       brand  \
count        50000        50000.000000     45518       50000   
unique          13                 NaN         7          40   
top      150,000km                 NaN    benzin  volkswagen   
freq         32424                 NaN     30107       10687   
mean           NaN            5.723360       NaN         NaN   
std            NaN            3.711984       NaN         NaN   
min            NaN            0.000000       NaN         NaN   
25%            NaN            3.000000       NaN         NaN   
50%            NaN            6.000000       NaN         NaN   
75%            NaN            9.000000       NaN         NaN   
max            NaN           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              40171                50000         50000.0  50000.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               35232                 1946             NaN           NaN   
mean                 NaN                  NaN             0.0  50813.627300   
std                  NaN                  NaN             0.0  25779.747957   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30451.000000   
50%                  NaN                  NaN             0.0  49577.000000   
75%                  NaN                  NaN             0.0  71540.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

- Any columns that have mostly one value that are candidates to be dropped
- Any columns that need more investigation.
- Any examples of numeric data stored as text that needs to be cleaned.

price and odometer_km columns are numeric values stored as text. Hence, remove any non-numeric characters then convert the values into numeric dtype.

replace() method is can only be used with string value, therefore convert the values into string first then replace the non-numeric characters or letter into "".

In [8]:
autos['price'] = autos['price'].str.replace("$", "")
autos['price'] = autos['price'].str.replace(",", "")
autos['odometer_km'] = autos['odometer_km'].str.replace("km", "")
autos['odometer_km'] = autos['odometer_km'].str.replace(",", "")


Convert dtype into numeric dtype since non-numeric characters are now removed.

In [9]:
autos['price'] = autos['price'].astype('int64')
autos['odometer_km'] = autos['odometer_km'].astype('int64')

---
## Data Preprocessing: Removing rows with outliers
Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.

In [10]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registrationn_year  \
0  privat    Angebot   5000  control          bus                2004   
1  privat    Angebot   8500  control    limousine                1997   
2  privat    Angebot   8990     test    limousine                2009   
3  privat    Angebot   4350  control   kleinwagen                2007   
4  privat    Angebot   1350     test        kombi                2003   

     gearbox  powerPS   model  odometer_km  registration_month fuel_type  \
0    manuell      158  andere       150000                   3       lpg   
1  automatik      286     7er       150000                   6    benzin   
2    manuell      102    golf        70000                   7    benzin   
3  automatik       71  fortwo        70000                   6    benzin   
4    manuell        0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### 'price' column

In [11]:
autos['price'].nunique()

2357

In [12]:
import seaborn as sns
import matplotlib.pyplot as plt

sns.boxplot(x=autos['price'])

plt.ticklabel_format(style='plain', axis='x',useOffset=False)

As shown in the boxplot above, there are outliers in 'price' column.

In [13]:
autos['price'].describe().apply(lambda x: format(x, 'f'))

count       50000.000000
mean         9840.043760
std        481104.380500
min             0.000000
25%          1100.000000
50%          2950.000000
75%          7200.000000
max      99999999.000000
Name: price, dtype: object

Mean value is approximately 9840 and Standard Deviation is 481104 therefore the values are dispersed by roughly 481104 on average. So I will consider the values larger than 9840 + 481104 = 490,944 as outliers.

In [14]:
autos['price'].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64

In [15]:
autos['price'].sort_index(ascending=False)

49999     1250
49998    22900
49997    13200
49996     1980
49995    24900
         ...  
4         1350
3         4350
2         8990
1         8500
0         5000
Name: price, Length: 50000, dtype: int64

In [16]:
autos['price'].sort_values(ascending=True).tail(20)

47337      259000
12682      265000
35923      295000
34723      299000
14715      345000
36818      350000
37585      999990
514        999999
43049      999999
22947     1234566
7814      1300000
47634     3890000
11137    10000000
24384    11111111
2897     11111111
27371    12345678
47598    12345678
39377    12345678
42221    27322222
39705    99999999
Name: price, dtype: int64

In [17]:
autos['price'][autos['price'] > 481104]

514        999999
2897     11111111
7814      1300000
11137    10000000
22947     1234566
24384    11111111
27371    12345678
37585      999990
39377    12345678
39705    99999999
42221    27322222
43049      999999
47598    12345678
47634     3890000
Name: price, dtype: int64

--> There are 14 outliers in 'price' column. Therefore, remove the rows with outliers as below.

In [18]:
autos = autos[autos['price'] < 481104]

--> 14 rows with outliers are removed from 'autos' dataframe.

In [19]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registrationn_year  \
0  privat    Angebot   5000  control          bus                2004   
1  privat    Angebot   8500  control    limousine                1997   
2  privat    Angebot   8990     test    limousine                2009   
3  privat    Angebot   4350  control   kleinwagen                2007   
4  privat    Angebot   1350     test        kombi                2003   

     gearbox  powerPS   model  odometer_km  registration_month fuel_type  \
0    manuell      158  andere       150000                   3       lpg   
1  automatik      286     7er       150000                   6    benzin   
2    manuell      102    golf        70000                   7    benzin   
3  automatik       71  fortwo        70000                   6    benzin   
4    manuell        0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

---
There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation

In [23]:
# normalize shows the percentage of each value takes
autos['date_crawled'].value_counts(normalize=True, dropna=False)

2016-03-14 20:50:02    0.00006
2016-03-29 23:42:13    0.00006
2016-03-09 11:54:38    0.00006
2016-03-08 10:40:35    0.00006
2016-04-02 15:49:30    0.00006
                        ...   
2016-03-28 21:58:22    0.00002
2016-03-14 09:36:23    0.00002
2016-03-24 14:59:02    0.00002
2016-03-21 07:37:06    0.00002
2016-03-22 13:56:22    0.00002
Name: date_crawled, Length: 48200, dtype: float64

To rank by date in ascending order

In [24]:
autos['date_crawled'].sort_index()

0        2016-03-26 17:47:46
1        2016-04-04 13:38:56
2        2016-03-26 18:57:24
3        2016-03-12 16:58:10
4        2016-04-01 14:38:50
                ...         
49995    2016-03-27 14:38:19
49996    2016-03-28 10:50:25
49997    2016-04-02 14:44:48
49998    2016-03-08 19:25:42
49999    2016-03-14 00:42:12
Name: date_crawled, Length: 49986, dtype: object

In [25]:
autos['date_crawled'].describe()

count                   49986
unique                  48200
top       2016-03-14 20:50:02
freq                        3
Name: date_crawled, dtype: object